packages/xhr-http-handler/src/xhr-http-handler.ts

```diff
@@ -0,0 +1,222 @@
+import { HttpHandler, HttpRequest, HttpResponse } from "@aws-sdk/protocol-http";
+import { buildQueryString } from "@aws-sdk/querystring-builder";
+import { HttpHandlerOptions, Provider } from "@aws-sdk/types";
+import { EventEmitter } from "events";
+
+import { requestTimeout } from "./request-timeout";
+
+/**
+ * Represents the http options that can be passed to the xhr http client.
+ */
+export interface XhrHttpHandlerOptions {
+  /**
+   * The number of milliseconds a request can take before being automatically
+   * terminated.
+   */
+  requestTimeout?: number;
+
+  /**
+   * Optionally way to provide instances of xhr.
+   * The default is `new XMLHttpRequest()` from the global scope.
+   * This can be used to attach listeners or other customization.
+   */
+  xhrFactory?: () => XMLHttpRequest | Promise<XMLHttpRequest>;
+
+  /**
+   * Will be called with the xhr instance before calling xhr.send.
+   * Can be used to attach additional event listeners.
+   */
+  xhrBeforeSend?: (xhr: XMLHttpRequest) => void;
+}
+
+/**
+ * An implementation of HttpHandler that uses XMLHttpRequest, which is
+ * traditionally associated with browsers.
+ */
+export class XhrHttpHandler extends EventEmitter implements HttpHandler {
+  /**
+   * Events emitted as an EventEmitter.
+   */
+  public static EVENTS = {
+    /**
+     * Emitted for xhr on progress.
+     * Payload is the native ProgressEvent.
+     */
+    PROGRESS: "progress",
+    /**
+     * Emitted for xhr.upload on progress.
+     * Payload is the native ProgressEvent.
+     */
+    UPLOAD_PROGRESS: "upload.progress",
+  };
+
+  private config?: XhrHttpHandlerOptions;
+  private readonly configProvider?: Provider<XhrHttpHandlerOptions>;
+
+  public constructor(options?: XhrHttpHandlerOptions | Provider<XhrHttpHandlerOptions | undefined>) {
+    super();
+    if (typeof options === "function") {
+      this.configProvider = async () => (await options()) || {};
+    } else {
+      this.config = options ?? {};
+    }
+  }
+
+  public destroy(): void {
+    // Do nothing. Connection pooling is handled by the browser.
+  }
+
+  public async handle(
+    request: HttpRequest,
+    { abortSignal }: HttpHandlerOptions = {}
+  ): Promise<{ response: HttpResponse }> {
+    if (!this.config && this.configProvider) {
+      this.config = await this.configProvider();
```

Comment:
Can we make sure the function is only called once at the concurrent invocation?

Reply:
will use the mechanism from node-http-handler. Should I update fetch-http-handler too?

Reply:
Yes, I think we missed it previously.